unzip_matrix_file <- function(file, unzip_func) {
  # Get the file extension preceding the .gz or .bz2 extension.
  unzipped_ext <- sub(".*\\.(.*)\\.(gz|bz2)", "\\1", file)
  # Special case for .tgz and .tbz2 files.
  if (grepl(".*\\.(tgz|tbz2)", file)) {
    unzipped_ext <- "tar"
  }
  # Unzip the compressed file.
  output_file <- tempfile(fileext = paste0(".", unzipped_ext))
  unzip_func(file, output_file, remove = FALSE)
  # Decompress tarballs in a temporary directory.
  if (unzipped_ext == "tar") {
    dir_suffix <- paste0(sample(letters, 10), collapse = "")
    temp_dir <- file.path(tempdir(), dir_suffix)
    dir.create(temp_dir, recursive = TRUE)
    utils::untar(output_file, exdir = temp_dir)
    output_file <- list.files(temp_dir, full.names = TRUE)[1]
  }
  output_file
}

#' Unzip (and untar) a .(t)gz or .(t)bz2 sparse matrix file
#' @note This function is a no-op for all other file extensions.
#'
#' @param file character(1) path to a potentially compressed sparse matrix file.
#'
#' @return character(1) path to unzipped file contents.
#' 
#' @import R.utils
#' @import tools
#' @noRd
get_decompressed_matrix_file <- function(file) {
  output_file <- file
  compressed_file_ext <- tolower(tools::file_ext(file))
  if (compressed_file_ext == "gz" || compressed_file_ext == "tgz") {
    output_file <- unzip_matrix_file(file, R.utils::gunzip)
  } else if (compressed_file_ext == "bz2" || compressed_file_ext == "tbz2") {
    output_file <- unzip_matrix_file(file, R.utils::bunzip2)
  }
  output_file
}

#' Load data from a 10X Genomics experiment
#'
#' Creates a \linkS4class{SparseMatrix} from the CellRanger output directories
#' for 10X Genomics data.
#'
#' @param sample character(1) directory name corresponding to a 10X sample. The
#'   directory should contain a matrix file, a gene/feature annotation file, and
#'   a barcode annotation file.
#'
#'   Alternatively, the string may contain a path to a HDF5 file in the sparse
#'   matrix format generated by 10X.
#'
#'   Alternatively, the string may contain a prefix of names for the three-file
#'   system described above, where the rest of the name of each file follows the
#'   standard 10X output.
#' @param col.names logical(1) indicating whether the columns of the matrix
#'   should be named with the cell barcodes.
#' @param row.names character(1) specifying whether to use Ensembl IDs ("id") or
#'   gene symbols ("symbol") as row names. If using symbols, the Ensembl ID will
#'   be appended to disambiguate in case the same symbol corresponds to multiple
#'   Ensembl IDs.
#' @param genome character(1) specifying the genome for HDF5 files output by
#'   CellRanger v2.
#' @param representation character(1) specifying the internal SparseMatrix
#'   representation (Default "svt").
#'
#' @return A \linkS4class{SparseMatrix} object containing count data for each
#'   gene (row) and cell (column) in \code{sample}.
#'
#' @details The signature of this function and its corresponding documentation
#' has largely been adapted from the \code{Read10xCounts} function in the
#' \pkg{DropletUtils} package.
#'
#' Note that user-level manipulation of sparse matrices requires loading of the
#' \pkg{SparseArray} package. Otherwise, calculation of \code{rowSums},
#' \code{colSums}, etc. will result in errors.
#'
#' @import Rcpp
#' @import SparseArray
#'
#' @examples
#' data("tenx_subset")  # Original dataset
#' new <- read_sparse_matrix(system.file(
#'   "extdata/tenx_subset.csv.gz", package = "smallcount"))
#' identical(new, tenx_subset)
#'
#' @references Zheng GX, Terry JM, Belgrader P, and others (2017). Massively
#' parallel digital transcriptional profiling of single cells. \emph{Nat Commun}
#' 8:14049.
#'
#' 10X Genomics (2017). Gene-Barcode Matrices.
#' \url{https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/output/matrices}
#'
#' 10X Genomics (2018). Feature-Barcode Matrices.
#' \url{https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices}
#'
#' 10X Genomics (2018). HDF5 Gene-Barcode Matrix Format.
#' \url{https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/2.2/advanced/h5_matrices}
#'
#' 10X Genomics (2018). HDF5 Feature Barcode Matrix Format.
#' \url{https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices}
#'
#' @export
#' @useDynLib smallcount
read_sparse_matrix <- function(sample,
                               col.names = FALSE,
                               row.names = c("id", "symbol"),
                               genome = NULL,
                               representation = c("svt", "coo")) {
  file <- get_decompressed_matrix_file(sample)
  row.names <- match.arg(row.names)
  genome <- ifelse(is.null(genome), "", genome)
  representation <- match.arg(representation)
  cppReadSparseMatrix(file, col.names, row.names, genome, representation)
}
